lib/oli/authoring/editing/container_editor.ex

```diff
@@ -264,6 +266,102 @@ defmodule Oli.Authoring.Editing.ContainerEditor do
     end
   end
 
+  @doc """
+    Duplicates a page within its project.
+    A true deep-copy of all the content is made. Resources and Revisions are created instead of simply referenced.
+  """
+  def duplicate_page(
+        %Revision{} = container,
+        page_id,
+        %Author{} = author,
+        %Project{} = project
+      ) do
+    original_page =
+      Resources.get_revision!(page_id)
+      |> Map.from_struct()
+
+    new_page_attrs =
+      original_page
+      |> Map.delete(:slug)
+      |> Map.delete(:inserted_at)
+      |> Map.delete(:updated_at)
+      |> Map.delete(:updated_at)
+      |> Map.delete(:resource_id)
+      |> Map.delete(:resource)
+      |> Map.put(:title, "Copy of #{original_page.title}")
+      |> Map.put(:content, nil)
+
+    Repo.transaction(fn ->
+      with {:ok, created_revision} <- add_new(container, new_page_attrs, author, project),
+           {:ok, duplicated_content} <-
+             duplicate_content_activities(
+               original_page.content,
+               project.slug,
+               author
+             ),
+           {:ok, updated_revision} <-
+             Resources.update_revision(created_revision, %{content: duplicated_content}) do
+        updated_revision
+      else
+        {:error, e} -> Repo.rollback(e)
+      end
+    end)
+  end
+
+  defp duplicate_content_activities(content, project_slug, author) do
+    case deep_copy_activities(content["model"], project_slug, author) do
```

Comment:
Seems like we could just call `deep_copy_activities` inside the transaction and avoid one function definition, wdyt?

Reply:
I get your point, I'm updating with some changes to perform the `deep_copy_activities` call inside the transaction, the less verbose way I came up with was to add another line in the `with`, let me know if that makes sense. 
I could also switch `%{original_page.content | "model" =>` for a `Map.put`, just went with what looked shorter.

Reply:
Yeah that seems fine 👍🏼